curriculum/challenges/english/25-front-end-development/lab-book-organizer/67172b43f84bcd2dec238a3d.md

```diff
@@ -0,0 +1,163 @@
+
+2. Your `books` array should have a minimum of three objects.
+
+3. You should have a callback function named `sortByYear` that accepts two books as parameter for sorting the array.
+
+4. The `sortByYear` function should return `-1` if the `releaseYear` of the first book is smaller than that of the second book.
+
+5. The `sortByYear` function should return `1` if the `releaseYear` of the first book is bigger than that of the second book.
+
+6. The `sortByYear` function should return `0` if both `releaseYear` values are equal.
+
+7. You should filter out books written after a certain year such as 1950 from the `books` array and save the filtered array in a new array named `filteredBooks`.
+
+8. You should sort the books in the `filteredBooks` array according to their `releaseYear` in ascending order and save the sorted array in a new array named `filteredBooksSorted`.
+
+# --hints--
+
+You should have a function `sortByYear` in your code.
+
+```js
+assert.isFunction(sortByYear);
+```
+
+Your `sortByYear` function should take two parameters.
+
+```js
+assert.lengthOf(sortByYear, 2);
+```
+
+Your `sortByYear` function should return `-1` if the `releaseYear` of `book1` object is smaller than that of the `book2` object, `1` if the `releaseYear` of `book1` object is larger than that of the `book2` object, and `0` in all other scenarios.
+
+```js
+assert.equal(sortByYear({releaseYear: 1913}, {releaseYear: 1925}), -1);
+assert.equal(sortByYear({releaseYear: 1925}, {releaseYear: 1913}), 1);
+assert.equal(sortByYear({releaseYear: 1925}, {releaseYear: 1925}), 0);
+```
+
+You should have an array `books` in your code.
+
+```js
+assert.isArray(books)
+```
+
+Your `books` array should have at least three book objects.
+
+```js
+assert.isAtLeast(books.length, 3)
+```
+
+Your `books` array should contain objects each with a string `title`, another string `authorName`, and a number `releaseYear`.
+
+```js
+
+books.forEach((book) => {
+  assert.isObject(book);
+  assert.sameMembers(Object.keys(book), ["title", "authorName", "releaseYear"]);
+  assert.isString(book.title);
+  assert.isNotEmpty(book.title);
+  assert.isString(book.authorName);
+  assert.isNotEmpty(book.authorName);
+  assert.isNumber(book.releaseYear);
+})
+```
+
+You should have an array `filteredBooks` in your code.
+
+```js
+assert.isArray(filteredBooks)
+```
+
+You should have an array `filteredBooksSorted` in your code.
+
+```js
+assert.isArray(filteredBooksSorted)
+```
+
+The `filteredBooksSorted` array should have some of the books filtered out from it and not be empty.
+
+```js
+assert.isBelow(filteredBooksSorted.length, books.length)
+assert.isAtLeast(books.length, 1)
+```
+
+You should call the `sort` higher order function by passing the `sortByYear` callback function on the `filteredBooks` array.
+
+```js
+assert.isTrue(filteredBooksSorted.every((book, index, array) => index === 0 || book.releaseYear >= array[index - 1].releaseYear))
```

Comment:
```suggestion
assert.match(__helpers.removeJSComments(code), //(const|let)\s+filteredBooksSorted\s*=\s*filteredBooks\s*\.\s*sort\s*\(\s*sortByYear\s*\)//)
```
What if we regex this?

Reply:
I agree with what we said in the meeting, but what's the point of having that callback function if you don't use it with `sort`? Also, this should be the higher order function lab, so I would expect at least one test checking that a higher order function is actually used somewhere.